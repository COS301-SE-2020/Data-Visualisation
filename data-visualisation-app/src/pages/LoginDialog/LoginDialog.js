/**
 *   @file LoginDialog.js
 *   Project: Data Visualisation Generator
 *   Copyright: Open Source
 *   Organisation: Doofenshmirtz Evil Incorporated
 *
 *   Update History:
 *   Date        Author              Changes
 *   -------------------------------------------------------
 *   14/7/2020   Byron Tominson      Original
 *
 *   Test Cases: data-visualisation-app/src/tests/LoginDialog.test.js
 *
 *   Functional Description:
 *   Provides a modal that promtes the user to login in or sign up.
 *
 *   Error Messages: "Error"
 *   Assumptions: None
 *   Constraints: None
 */

import React from 'react';
import {useState} from 'react';
import {Button, Modal, Input, Tooltip, AutoComplete, Select} from 'antd';
import {Form, Checkbox, Spin} from 'antd';
import { QuestionCircleOutlined } from '@ant-design/icons';
import request from '../../globals/requests';
import * as constants from '../../globals/constants';
import {useGlobalState} from '../../globals/Store';
import './LoginDialog.scss';
import { notification } from 'antd';



const { Option } = Select;
const AutoCompleteOption = AutoComplete.Option;

const formItemLayout = {
  labelCol: {
    xs: { span: 24 },
    sm: { span: 8 },
  },
  wrapperCol: {
    xs: { span: 24 },
    sm: { span: 16 },
  },
};
const tailFormItemLayout = {
  wrapperCol: {
    xs: {
      span: 24,
      offset: 0,
    },
    sm: {
      span: 16,
      offset: 8,
    },
  },
};

/**
  * Constants 
*/
const layout = {
  labelCol: { span: 8 },
  wrapperCol: { span: 16 },
};
const tailLayout = {
  wrapperCol: { offset: 8, span: 16 },
};



/**
  * @param props passed from LoginDialog function.
  * @return React Component
*/
function SignUpDialog(props) {

  const [confirmLoading, setConfirmLoading] = useState(false);

  /**
    * Notifications
  */
  const openNotification = placement => {
    notification.info({
      message: 'Invalid data',
      description:
        'The server rejected your infomation',
      placement,
    });
  };
  const signUpSuccessNotification = placement => {
    notification.info({
      message: 'Sign up success',
      description:
        'You are now successfully signed up with us',
      placement,
    });
  };


  const [form] = Form.useForm();


  /**
    * Send request to backend with form infomation.
  */
  const onFinish = values => {
    setConfirmLoading(true);
    request.user.register(values.name, values.surname, values.email, values.password, values.confirm, function(result) {
      console.log(result);
      if (result === constants.RESPONSE_CODES.SUCCESS) {
        /**
          * On Success
        */
        setConfirmLoading(false);
        setVisible(false);
        signUpSuccessNotification('bottomRight');
      }
      else{
        /**
          * On fail
        */
        setConfirmLoading(false);
        openNotification('bottomRight');
      }
    });
    
  };

  /**
    * Autocomplete
  */
  const [autoCompleteResult, setAutoCompleteResult] = useState([]);
  const websiteOptions = autoCompleteResult.map(website => ({
    label: website,
    value: website,
  }));

  const [visible, setVisible] = useState(true);

  /**
    * Actions for cancel
  */
  function handleCancel(e) {
      setVisible(false);
  };


  return (
    <div >
      <Modal
          title='Sign Up'
          visible={visible}
          onCancel={handleCancel}

          footer={[
            
            ]}
          >
        <Spin spinning={confirmLoading}></Spin>
        <Form
          {...formItemLayout}
          form={form}
          name="register"
          onFinish={onFinish}
          scrollToFirstError
        >
          <Form.Item
            name="name"
            label={
              <span>
                Name&nbsp;
                <Tooltip title="What do you want others to call you?">
                  <QuestionCircleOutlined />
                </Tooltip>
              </span>
            }
            rules={[{ required: true, message: 'Please input your name!', whitespace: true }]}
          >
            <Input />
          </Form.Item>
          <Form.Item
            name="surname"
            label='Surname'
            rules={[{ required: true, message: 'Please input your surname!', whitespace: true }]}
          >
            <Input />
          </Form.Item>

          <Form.Item
            name="email"
            label="E-mail"
            rules={[
              {
                type: 'email',
                message: 'The input is not valid E-mail!',
              },
              {
                required: true,
                message: 'Please input your E-mail!',
              },
            ]}
          >
            <Input />
          </Form.Item>

          <Form.Item
            name="password"
            label="Password"
            rules={[
              {
                required: true,
                message: 'Please input your password!',
              },
            ]}
            hasFeedback
          >
            <Input.Password />
          </Form.Item>

          <Form.Item
            name="confirm"
            label="Confirm Password"
            dependencies={['password']}
            hasFeedback
            rules={[
              {
                required: true,
                message: 'Please confirm your password!',
              },
              ({ getFieldValue }) => ({
                validator(rule, value) {
                  if (!value || getFieldValue('password') === value) {
                    return Promise.resolve();
                  }
                  return Promise.reject('The two passwords that you entered do not match!');
                },
              }),
            ]}
          >
            <Input.Password />
          </Form.Item>

          <Form.Item
            name="agreement"
            valuePropName="checked"
            rules={[
              { validator:(_, value) => value ? Promise.resolve() : Promise.reject('Should accept agreement') },
            ]}
            {...tailFormItemLayout}
          >
            <Checkbox>
              I have read the <a href="">agreement</a>
            </Checkbox>
          </Form.Item>
          <Form.Item {...tailFormItemLayout}>
            <Button type="primary" htmlType="submit">
              Register
            </Button>
          </Form.Item>
        </Form>
        <Spin spinning={confirmLoading}></Spin>
      </Modal>
      
    </div>
  
  );
}




/**
  * @param props passed from App function.
  * @return React Component
*/
function LoginDialog(props) {

  
  const [visible, setVisible] = useState(false);
  const [signup, setSignUp] = React.useState('false');
  const [confirmLoading, setConfirmLoading] = useState(false);
 
  
  /**
    * Notifications
  */
  const openNotification = placement => {
    notification.info({
      message: 'Invalid credentials',
      description:
        'Either your email or password was incorrect. Try again',
      placement,
    });
  };
  const loginSuccessNotification = placement => {
    notification.info({
      message: 'Welcome',
      description:
        'Hope you have a great day',
      placement,
    });
  };
  const logoutSuccessNotification = placement => {
    notification.info({
      message: 'Bye',
      description:
        'Have a good day',
      placement,
    });
  };

  /**
    * Make modal visible or not
  */
  function showModal() {
    setVisible(true);
    setSignUp(false);
  };

  /**
    * Actions for cancel
  */
  function handleCancel() {
      setVisible(false);
  };

  /**
    * Function to change signUp state to 'true'
  */
  function handleSignUp(e) {
    setVisible(false);
    setSignUp('true');
  };


  /**
    * useGlobalState used and deglared from Store.js
  */
  const [state, dispatch] = useGlobalState();
 

  /**
    * Logs the user out.
    * Sends the request to backend.
    * After success: 
    * Resets the data sources list
    * Resets the page type back to home
    * Resets the explore stage
    * Resets the dashbaord stage
    * Changes global isLoggedIn variable to false
    * Displays notification 
  */
  function handleLogout(){
    /**
     * Send request to backend
    */
    request.user.logout(function(result) {
      console.log(result);
      if (result === constants.RESPONSE_CODES.SUCCESS) {
        /**
          * Reset on success
        */
        request.user.dataSources = [
          {
            'id': 6,
            'email': 'elna@gmail.com',
            'sourceurl': 'https://services.odata.org/V2/Northwind/Northwind.svc'
          }
        ];
        props.handlePageType('home');
        props.setDashboardIndex('');
        props.setDashboardStage('dashboardHome');
        props.setIsAddingDashboard(false);
        props.setExploreStage('dataConnection');
        dispatch({ isLoggedIn: false }); 
        logoutSuccessNotification('bottomRight');
      }
    });
 }


  /**
    * Sends request to backend with form infomation.
  */   
  const onFinish = values => {
    setConfirmLoading(true);
    request.user.login(values.email, values.password, values.remember, function(result) {
      console.log(result);
      if (result === constants.RESPONSE_CODES.SUCCESS) {
          /**
            * On success
          */
          setConfirmLoading(false);
          dispatch({ isLoggedIn: true });
          setVisible(false);
          props.handlePageType('home');
          props.setExploreStage('dataConnection');
          loginSuccessNotification('bottomRight');
      }
      else{
        /**
          * On fail
        */
        setConfirmLoading(false);
        openNotification('bottomRight');
      }
    });
 };

  /**
    * Handle finish error.
  */  
  const onFinishFailed = errorInfo => {
    console.log('Failed:', errorInfo);
  };

  return (
    <div id = 'loginDiv'>
    
        {
          state.isLoggedIn === false ?
            <Button ghost className='button__login' id = 'loginButton' type="dashed" onClick={showModal}>Login/Sign Up</Button>
            :
            <Button ghost className='button__login' id = 'logout' type="dashed" onClick={handleLogout}>Logout</Button>
        }
      <Modal
          title="Login"
          visible={visible}
          onOk={onFinish}
          onCancel={handleCancel}
          confirmLoading={confirmLoading}
          footer={[
         
          ]}
          >
          <Spin spinning={confirmLoading}>
          <Form
            {...layout}
            name="basic"
            initialValues={{ remember: true }}
            onFinish={onFinish }
            onFinishFailed={onFinishFailed}
          >
            <Form.Item
              name="email"
              label="E-mail"
              rules={[
                {
                  type: 'email',
                  message: 'The input is not valid E-mail!',
                },
                {
                  required: true,
                  message: 'Please input your E-mail!',
                },
              ]}
            >
              <Input />
            </Form.Item>

            
            <Form.Item
              name="password"
              label="Password"
              rules={[{ required: true, message: 'Please input your password!' }]}
            >
              <Input.Password />
            </Form.Item>

            <Form.Item {...tailLayout} name="remember" valuePropName="checked">
              <Checkbox>Remember me</Checkbox>
            </Form.Item>

            <Form.Item {...tailLayout}>
          

              <Button type="primary" htmlType="submit">
                Login
              </Button>

              <Button htmlType="button" style={{marginLeft: '10px'}} onClick = {handleSignUp}>
                Sign Up
              </Button>
              
            </Form.Item>

          </Form>
          </Spin>
      </Modal>
     
      
      <main>
        {
          signup === 'true' ?
            <SignUpDialog/>
            :
            null
        }
      </main>
    </div>
    
    
  );
}

export default LoginDialog;