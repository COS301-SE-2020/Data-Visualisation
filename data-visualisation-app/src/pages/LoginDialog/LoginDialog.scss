
#login-div{  
    left:81%;
    position: absolute;
    display:inline-block;
    overflow: auto;
    margin:0px auto;
    border-radius: 25px;
    width: 100px;
    width: 270px;
}

@media only screen and (max-width: 1660px) {
    #login-div{
        left: 77%;
    }
}

@media only screen and (max-width: 1410px) {
    #login-div{
        left: 71%;
    }
}

@media only screen and (max-width: 1170px) {
    #login-div{
        left: 65%;
    }
}

@media only screen and (max-width: 1030px) {
    #login-div{
        left: 60%;
    }
}

@media only screen and (max-width: 920px) {
    #login-div{
        left: 55%;
    }
}

@media only screen and (max-width: 845px) {
    #login-div{
        left: 50%;
    }
}

@media only screen and (max-width: 780px) {
    #login-div{
        left: 40%;
    }
}

@media only screen and (max-width: 690px) {
    #login-div{
        left: 35%;
    }
}

@media only screen and (max-width: 650px) {
    #login-div{
        left: 30%;
    }
}

@media only screen and (max-width: 600px) {
    #login-div{
        left: 50%;
    }
}

@media only screen and (max-width: 540px) {
    #login-div{
        left: 40%;
    }
}

@media only screen and (max-width: 450px) {
    #login-div{
        left: 30%;
    }
}


@media only screen and (max-width: 380px) {
    #login-div{
        left: 22%;
    }
}


@media only screen and (max-width: 320px) {
    #login-div{
        left: 10%;
        
    }
    .dropdown-btn > .ant-btn-default  {
        font-size: 7pt;
    }
}

@media only screen and (max-width: 280px) {
    #login-div{
        left: 4%;
        
    }
    .dropdown-btn > .ant-btn-default  {
        font-size: 5.5pt;
    }
}








#loginButton{
    background-color: #242424;
    color: #3EC195;
    border: none;
    font-weight: 900;
    float: right;
}

.dropdown-btn > .ant-dropdown-trigger  {
    background-color: #242424;
    color: #3EC195;
    border: none;
}

.dropdown-btn > .ant-btn-default  {
    background-color: #242424;
    color: #3EC195;
    border: none;
    border-radius: 25px;
    font-weight: 900;
}

.dropdown-btn > .ant-btn-default:hover  {
    background-color: #242424;
    color: #3EC195;
  
}

.dropdown-btn > .ant-btn-default:focus  {
    background-color: #242424;
    color: #3EC195;
}

.dropdown-btn{
    float: right;
}
