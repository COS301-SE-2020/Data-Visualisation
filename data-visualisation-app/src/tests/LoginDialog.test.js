import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import LoginDialog from '../pages/LoginDialog/LoginDialog';
import renderer from 'react-test-renderer';


describe('LoginDialog component renders correctly', () => {
    it('renders correctly', () => {
      const rendered = renderer.create(<LoginDialog/>);
      expect(rendered.toJSON()).toMatchSnapshot();
    });
});


// describe('Input value', () => {

//     // it('updates on change', () =>{
//     //     const {queryByPlaceholderText} = render(<AddDashboard/>);

//     //     const addInput = queryByPlaceholderText('Consumer');

//     //     fireEvent.change(addInput, {target: {value: 'test'}});

//     //     expect(addInput.value).toBe('test');
//     // });
// });



