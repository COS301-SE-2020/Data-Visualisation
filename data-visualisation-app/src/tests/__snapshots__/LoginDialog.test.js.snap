// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDialog component renders correctly renders correctly 1`] = `
<div
  id="loginDiv"
>
  <button
    className="ant-btn ant-btn-dashed"
    id="loginButton"
    onClick={[Function]}
    style={
      Object {
        "color": "#3C6A7F",
      }
    }
    type="button"
  >
    <span>
      Login/Sign Up
    </span>
  </button>
  <main />
</div>
`;
